### PR TITLE
490 check operator SOLs

### LINKED ISSUE #490: Check available operator SOLs before executing Neon transaction

### PR DESCRIPTION
**Pull request by vasiliy-zaznobin**
If the balance value of one of the operator's accounts has become equal to
        the value of the variable MIN_OPERATOR_BALANCE_TO_ERR or less,
        then an error is written to the log and the next operator's accounts is used (and check balance again):
```
ERROR:neon.Proxy:Operator account BMp6gEnveANdvSvspESJUrNczuHz1GF5UQKjVLCkAZih has NOT enough SOLs; balance = 7976686866080; min_operator_balance_to_err = 1049000000000000000000000000
```
----
If the balance value of one of the operator's accounts becomes equal to
        the value of the variable MIN_OPERATOR_BALANCE_TO_WARN or less,
        then a warning is written to the log:
```
WARNING:neon.Proxy:Operator account BMp6gEnveANdvSvspESJUrNczuHz1GF5UQKjVLCkAZih SOLs are running out; balance = 7976686866080; min_operator_balance_to_warn = 1049000000000000000000000000; min_operator_balance_to_err = 1049049000; 
```
----
If the balance value of the all operator's accounts has become equal to
        the value of the variable MIN_OPERATOR_BALANCE_TO_ERR or less,
        then an error (RuntimeError('No resources!') )is returned to the client
        who requested the execution of the transaction
        and an error is written to the log.

### CODE REVIEW

**Review by afalaleev — APPROVED** [review]

File: proxy/common_neon/transaction_sender.py [hunk 1]
```diff
@@ -412,13 +414,31 @@ def clear_resource(self):
...
+        return MIN_OPERATOR_BALANCE_TO_WARN
+
     def _validate_execution(self):
         # Validate that operator has available resources: operator key, holder/storage accounts
         self._resource_list.init_resource_info()
 
         self._validate_pend_tx()
         self._validate_whitelist()
         self._validate_tx_count()
+        self._validate_operator_balance()
```
Comment (afalaleev) [inline proxy/common_neon/transaction_sender.py]: In this place the operator key is already selected and the transaction execution fails as result.
Better to add this validation to https://github.com/neonlabsorg/proxy-model.py/blob/d6ff1e0f8bda95b76776d53886be360ab0a96fe3/proxy/common_neon/transaction_sender.py#L300-L302

As result, the OperatorResourceList will switch to another operator key.
Fail happens only when OperatorResourceList doesn't have operator key with enough SOLs.

**Review by afalaleev — CHANGES_REQUESTED** [review]

File: proxy/common_neon/transaction_sender.py (hunk 1 above)
Comment (vasiliy-zaznobin) [reply proxy/common_neon/transaction_sender.py]: If it was not possible to select/create a resource, this is a different situation and we do not handle it here.
The resource creation process should have taken care of the erroneous situations. If this is not the case, then this is a mistake (in this process) and should be done in a separate issue.

Comment (vasiliy-zaznobin) [reply proxy/common_neon/transaction_sender.py]: The new behavior has been implemented

File: proxy/common_neon/transaction_sender.py [hunk 2]
```diff
@@ -255,7 +255,7 @@ def _init_resource_list(self):
             if self._resource_list_len_glob.value == 0:
                 raise RuntimeError('No resources!')
 
-    def init_resource_info(self) -> OperatorResourceInfo:
+    def init_resource_info(self, solana: SolanaInteractor) -> OperatorResourceInfo:
```
Comment (afalaleev) [inline proxy/common_neon/transaction_sender.py]: This parameter is surplus.
OperatorResourceList already has solana: `self._s.solana`

File: proxy/common_neon/transaction_sender.py [hunk 3]
```diff
@@ -270,14 +270,15 @@ def init_resource_info(self) -> OperatorResourceInfo:
 
             with self._resource_list_len_glob.get_lock():
                 if self._resource_list_len_glob.value == 0:
+                    self.error(f'Operator has NO resources!')
```
Comment (afalaleev) [inline proxy/common_neon/transaction_sender.py]: This log record is surplus

All exceptions with tracebacks are printed in this place - https://github.com/neonlabsorg/proxy-model.py/blob/0e9ada600ae723ff6ee9512fafc8fbbadf77dded/proxy/plugin/solana_rest_api.py#L505-L507

**Review by afalaleev — CHANGES_REQUESTED** [review]

File: proxy/common_neon/transaction_sender.py (hunk 2 above)
Comment (vasiliy-zaznobin) [reply proxy/common_neon/transaction_sender.py]: Fixed

File: proxy/common_neon/transaction_sender.py (hunk 3 above)
Comment (vasiliy-zaznobin) [reply proxy/common_neon/transaction_sender.py]: Fixed

**Review by afalaleev — APPROVED** [review]

**Pull request merged**